### PR TITLE
[tutorials][ML] Add RDataLoader classification tutorial with XGBoost

### PR DESCRIPTION
**Pull request by siliataider**
# This Pull request:

## Changes:
Adds a new tutorial `ml_dataloader_XGBoost.py` demonstrating XGBoost training using RDataLoader to stream data directly from EOS without intermediate files stored after the shuffling/splitting phase.

This tutorial is inspired from the [tmva101_Training](https://root.cern/doc/master/tmva101__Training_8py.html) workflow but eliminates the need for the [tmva100_DataPreparation](https://root.cern/doc/master/tmva100__DataPreparation_8py.html) step which splits the dataset in 2 separate training/testing sets and stores 2 intermediate files.

### CODE REVIEW

**Review by guitargeek — COMMENTED** [review]
I'm curious: is there an advantage of using the batching data loader for loading all data at once? `RDataFrame.AsNumpy` seems completely fine here.

Your new code can maybe be seen as a usage example of the `RDataLoader`, but I would rather do this only in tutorials where the data is actually loaded in batches. Otherwise, we risk sending the wrong message that getting your ROOT data to Numpy is not as easy as one line `RDataFrame.AsNumpy`.

**Comment by siliataider** [comment]
Fair point.

There are a couple of advantages of using the rdataloader though:
- you can define your label and wieght columns with rdataframe and unpack them with the rdataloader as the target/weights separation is built-in instead of manually creating separate arrays
- using the rdataloader essentially makes the entire [data preparation](https://root.cern/doc/master/tmva100__DataPreparation_8py.html) script unnecessary as you can do both shuffling and splitting dynamically at train time

So:
- either i merge the `tmva100_DataPreparation.py` and `tmva101_Training.py` because using the `RDataLoader` we don't need both,
- or I create a separate example inspired from this one, just with this alternative laoding mechanism

My goal at the end is to have a tutorial showing how to do XGBoost training with the RDataLoader

**Comment by guitargeek** [comment]
> Fair point.
> 
> There are a couple of advantages of using the rdataloader though:
> 
>     * you can define your label and wieght columns with rdataframe and unpack them with the rdataloader as the target/weights separation is built-in instead of manually creating separate arrays

As you see in the existing tutorial, adding the weights as NumPy arrays is not hard, so I wouldn't really call it an advantage of the data loader.

> 
>     * using the rdataloader essentially makes the entire [data preparation](https://root.cern/doc/master/tmva100__DataPreparation_8py.html) script unnecessary as you can do both shuffling and splitting dynamically at train time
> 
> 
> So:
> 
>     * either i merge the `tmva100_DataPreparation.py` and `tmva101_Training.py` because using the `RDataLoader` we don't need both,

The same argument I made earlier still applies: even if you merge the `tmva100_DataPreparation` and `tmva101_Training`, you could just do a `AsNumpy` in the end, so I don't think this justifies the RDataLoader here.

> 
>     * or I create a separate example inspired from this one, just with this alternative laoding mechanism
> 
> 
> My goal at the end is to have a tutorial showing how to do XGBoost training with the RDataLoader

That's not what you show in this tutorial. You're showing how to convert ROOT data to NumPy arrays via the RDataLoader, for which I don't see why you'd do that :slightly_smiling_face: And now the tutorial sends the (what I think wrong) message that the RDataLoader is the right way to do this conversion. 

And since the XGBoost algo is also not designed to re-read training data in each iteration, but it converts it into an internal quantized representation once in the beginning, there is also no way do directly hook in the data loader into XGBoost training.

**Comment by siliataider** [comment]
Fair enough modifying `tmva101` wasn't the right approach probably, I've created a separate tutorial instead: `ml_dataloader_XGBoost.py`

The use case comes from workflows where users want to:
- Retrain with different shuffling seeds without regenerating intermediate files
- Stream directly from EOS without local `Snapshot` copies of subsets of the original dataset

You're correct that for simple one-shot loading, `AsNumpy()` is sufficient and simpler, however if users want to do more, like shuffling, splitting datasets without saving intermediate files, under/oversampling etc. the `RDataLoader` is a more convenient choice, this tutorial shows that

File: tutorials/machine_learning/ml_dataloader_XGBoost.py [hunk 1]
```diff
@@ -0,0 +1,96 @@
+## \file
+## \ingroup tutorial_ml
+## \notebook -nodraw
+## This tutorial demonstrates XGBoost training using RDataLoader to load data
+## directly from ROOT files without intermediate preparation steps.
+## This is an alternative implementation to tutorial tmva101_Training.py,
+## which stores intermediate files after train/test splitting
+## and uses numpy arrays to load data.
```
Comment (guitargeek) [inline tutorials/machine_learning/ml_dataloader_XGBoost.py]: I would not mention `tmva101`. Cross-referencing between tutorials has a risk to not age well, in particular here because we definitely have to do something about `tmva101`: it encourages users to store experimental objects like `TMVA::Experimental::RBDT` in a ROOT file via `ROOT.TMVA.Experimental.SaveXGBoost`, without explicitly telling you to **not** do this for production purposes. Yes, the `SaveXGBoost` *interface* is clearly experimental, but it doesn't warn the users that the content of the generated file will also be experimental classes.

File: tutorials/machine_learning/index.md [hunk 2]
```diff
@@ -137,5 +137,6 @@
 | ml_dataloader_PyTorch.py | Loading batches of events from a ROOT dataset into a basic PyTorch workflow. |
 | ml_dataloader_TensorFlow.py | Loading batches of events from a ROOT dataset into a basic TensorFlow workflow. |
 | ml_dataloader_Higgs_Classification.py | Loading batches of events from different files for a data-normalization workflow. |
-| ml_dataloader_resampling.py | Loading batches of events from an imbalanced ROOT dataset and balancing them. | 
+| ml_dataloader_resampling.py | Loading batches of events from an imbalanced ROOT dataset and balancing them. |
+| ml_dataloader_XGBoost.py | Training XGBoost directly from remote ROOT files. |
```
Comment (guitargeek) [inline tutorials/machine_learning/index.md]:
```suggestion
| ml_dataloader_XGBoost.py | Training classifier models directly from remote ROOT files, shown with an XGBoost example. |
```
The pattern you show generalized very well to other classification algos, so I would make that clear before people ask you to re-write the same tutorial for catboost or sckit-learn algorithms

**Review by guitargeek — APPROVED** [review]
Nice, thank you! Just two small suggestions: I'd refrain from mentioning `tmva101` and make it more clear that this tutorial is not XGBoost specific: you'd do exactly the same for any other classifier. Maybe even show a different one, like [logistic regression from scikit learn](https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LogisticRegression.html). I'd even consider renaming the tutorial to `ml_dataloader_classification.py`, but actually sticking with XGBoost in the title might make it more attractive... Can't decide for myself :slightly_smiling_face:

File: tutorials/machine_learning/ml_dataloader_XGBoost.py [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+
+
+def load_data():
+    """Load signal and background data"""
+    rdf_sig = prepare_rdf("SMHiggsToZZTo4L.root", 1)
+    rdf_bkg = prepare_rdf("ZZTo2e2mu.root", 0)
+
+    # Compute class-balancing weights
+    num_sig = rdf_sig.Count().GetValue()
+    num_bkg = rdf_bkg.Count().GetValue()
```
Comment (vepadulano) [inline tutorials/machine_learning/ml_dataloader_XGBoost.py:52]: micro-optimization here: given there are two distinct computation graphs, `RunGraphs` could be used to compute the number of entries of both in parallel. Although, `EnableImplicitMT` is not called in this tutorial so we may want to avoid doing that in order not to mix topics.

File: tutorials/machine_learning/ml_dataloader_XGBoost.py [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+
+    rdf_sig = rdf_sig.Define("weight", f"{num_all}.0/{num_sig}.0")
+    rdf_bkg = rdf_bkg.Define("weight", f"{num_all}.0/{num_bkg}.0")
+
+    loader = ROOT.Experimental.ML.RDataLoader(
+        [rdf_sig, rdf_bkg],
+        columns=variables + ["label", "weight"],
+        target="label",
+        weights="weight",
+        batch_size=num_all,  # Load all data in one batch
```
Comment (vepadulano) [inline tutorials/machine_learning/ml_dataloader_XGBoost.py:63]: Can you remind me why we don't do directly `eager=True` in this case?

File: tutorials/machine_learning/ml_dataloader_XGBoost.py [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+        drop_remainder=False,
+        set_seed=42,
+    )
+
+    # Split into training and testing sets
+    train, test = loader.train_test_split(test_size=0.5)
+
+    X_train, y_train, w_train = next(iter(train.as_numpy()))
+    X_test, y_test, w_test = next(iter(test.as_numpy()))
+
```
Comment (vepadulano) [inline tutorials/machine_learning/ml_dataloader_XGBoost.py:76]: This is where the iterator is actually loaded into a single object(s) in memory, maybe here is where we should have a comment.

**Review by vepadulano — APPROVED** [review]
Thank you! A couple of minor considerations from my side

File: tutorials/machine_learning/ml_dataloader_XGBoost.py (hunk 4 above)
Comment (siliataider) [reply tutorials/machine_learning/ml_dataloader_XGBoost.py:63]: With `batch_size=dataset_size` lazy and eager loading are functionally equivalent here as we're loading everything in a single batch anyway, the main reason we don't use `load_eager=True` is that it doesn't yet support passing multiple dataframes (outside of resampling), and we need to pass both `rdf_sig` and `rdf_bkg` here ( + it's an extra flag to set)
